### PR TITLE
Livechat queue methods options & Guest Pool queue method

### LINKED ISSUE #3433: Routing of new livechat to agents Algorithms
Mentioned in #1824. Admins should have the ability to pick from multiple new chat to agents routing algorithms. Currently the agent with the lowest total conversations is paired with the new chat. Other possibilities are random pairing, pairing in a specific "batting" order, or the method bellow. 

Another method is allowing agents to pick from a pool of open chats yet to be respond to.  During this time the client would could be notified that they are "currently trying to connect to an agent". 
screen shot 2016-06-01 at 3 27 21 pm
screen shot 2016-06-01 at 3 27 32 pm
screen shot 2016-06-01 at 3 27 44 pm

This algorithm does cause some difficulties though... 
- One solution: add all agents to the new room, and remove all but the one who takes it when they take it. However that would mean a room would need some sort of "taken" property, and the client side would need some sort of "takeRoom" method which removes everyone from except for themselves from the room. One problem with this is anybody could remove everyone except themselves from any livechat room if they had the room id..
- Another solution: introduce the concept an _inquiry_ (or whatever you call it), a separate data structure that has the new room id (with only the client in it) and some other info (timestamp, initial question, title, etc..). You show all agents all inquiries, and when they take one you add the agent to the corresponding room and remove the inquiry. This abstracts away from removing users from rooms, but introduces a new data structure. In addition it doesn't allow the agent to preview the room, though you could simply use a popup as demonstrated above and limit the client to one message until they are connected. 

This feature overall is a pretty unique from the other routing algorithms and might be a lot to turn on or off.

### PR DESCRIPTION
**Pull request by AlecTroemel**
@RocketChat/core 

Closes #3433 

Added option to choose between multiple live chat queuing methods. The old way is now the "least amount" 

Also added another method called "Guest Pool" where agents take clients from open inquiries within your departments. 
screen shot 2016-06-09 at 12 35 52 pm
screen shot 2016-06-09 at 12 36 48 pm
screen shot 2016-06-09 at 12 37 03 pm
screen shot 2016-06-09 at 12 37 15 pm

### CODE REVIEW

File: packages/rocketchat-livechat/package.js [hunk 1]
```diff
@@ -139,14 +141,18 @@ Package.onUse(function(api) {
 	api.addFiles('server/models/LivechatPageVisited.js', 'server');
 	api.addFiles('server/models/LivechatTrigger.js', 'server');
 	api.addFiles('server/models/indexes.js', 'server');
+	api.addFiles('server/models/LivechatInquiry.js', 'server');
 
 	// server lib
 	api.addFiles('server/lib/Livechat.js', 'server');
+	api.addFiles('server/lib/QueueMethods.js', 'server');
+
```
Comment (rodrigok) [inline packages/rocketchat-livechat/package.js]: Codacy Issue found: [Trailing spaces not allowed.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2726404244/issues/source?bid=3199634&fileBranchId=3348698#l149)

File: packages/rocketchat-ui-sidenav/side-nav/livechatInquiryItem.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
+Template.livechatInquiryItem.helpers({
+    alert: function() {
+        if (FlowRouter.getParam('_id') !== this.rid || !document.hasFocus()) {
+            return this.alert;
+        }
+    },
+    unread: function() {
+        return 1;
+        if ((FlowRouter.getParam('_id') !== this.rid || !document.hasFocus()) && this.unread > 0) {
```
Comment (rodrigok) [inline packages/rocketchat-ui-sidenav/side-nav/livechatInquiryItem.js]: Codacy Issue found: [Found unexpected statement after a return.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2726404390/issues/source?bid=3199634&fileBranchId=3348698#l9)

File: packages/rocketchat-ui-sidenav/side-nav/livechatInquiryItem.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        var inquiry = this;
+
+        swal({
+            title: 'Do you want to take this client?',
+            text: 'message: ' + inquiry.message,
+            showCancelButton: true,
+            confirmButtonColor: '#3085d6',
+            cancelButtonColor: '#d33',
+            confirmButtonText: 'Take it!'
+        }, function (isConfirm) {
```
Comment (rodrigok) [inline packages/rocketchat-ui-sidenav/side-nav/livechatInquiryItem.js]: Codacy Issue found: [Unexpected space before function parentheses.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2726404390/issues/source?bid=3199634&fileBranchId=3348698#l52)

File: packages/rocketchat-ui-sidenav/side-nav/livechatInquiryItem.js [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+            if (isConfirm) {
+                // make server method call
```
Comment (rodrigok) [inline packages/rocketchat-ui-sidenav/side-nav/livechatInquiryItem.js]: Codacy Issue found: [Trailing spaces not allowed.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2726404390/issues/source?bid=3199634&fileBranchId=3348698#l54)

File: packages/rocketchat-livechat/server/models/LivechatInquiry.js [hunk 5]
```diff
@@ -0,0 +1,73 @@
+class LivechatInquiry extends RocketChat.models._Base {
+	constructor() {
+		super();
+		this._initModel('livechat_inquiry');
+
+		this.tryEnsureIndex({ 'rid': 1 }); // room id corresponding to this inquiry
```
Comment (rodrigok) [inline packages/rocketchat-livechat/server/models/LivechatInquiry.js]: Codacy Issue found: [Trailing spaces not allowed.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2726404280/issues/source?bid=3199634&fileBranchId=3348698#l6)

File: packages/rocketchat-livechat/server/models/LivechatInquiry.js [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+	// 	// remove agent subscription 
+	// 	RocketChat.models.Subscriptions.removeByRoomIdAndUserId(roomId, agentId);
+
+	// 	// remove agent username from room 
+	// 	RocketChat.models.Rooms.removeUsernameById(roomId, username);
+
+	// 	// mark inquiry as open 
+	// 	this.update({},{ 
+	// 		$set: { status: 'open' }
```
Comment (rodrigok) [inline packages/rocketchat-livechat/server/models/LivechatInquiry.js]: Codacy Issue found: [Trailing spaces not allowed.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2726404280/issues/source?bid=3199634&fileBranchId=3348698#l68)

File: packages/rocketchat-livechat/server/models/LivechatInquiry.js [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+		// add user to room 
+		RocketChat.models.Rooms.addUsernameById(inquiry.rid, agent.username);
+
+		// TODO: mark inquiry as taken 
+		this.update({},{ 
+			$set: { status: 'taken' } 
+		});
+
+		// add status field 
+		// this.remove({_id: inquiry._id});
```
Comment (rodrigok) [inline packages/rocketchat-livechat/server/models/LivechatInquiry.js]: Codacy Issue found: [Trailing spaces not allowed.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2726404280/issues/source?bid=3199634&fileBranchId=3348698#l51)

File: packages/rocketchat-livechat/server/models/LivechatInquiry.js [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+		// add status field 
+		// this.remove({_id: inquiry._id});	
+
+	}
+
+	// returnAsInquiry(roomId, agentId) {
+	// 	// find out corresponding inquiry from room, 
+
+	// 	//get agent username
```
Comment (rodrigok) [inline packages/rocketchat-livechat/server/models/LivechatInquiry.js]: Codacy Issue found: [Trailing spaces not allowed.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2726404280/issues/source?bid=3199634&fileBranchId=3348698#l58)

File: packages/rocketchat-livechat/server/models/LivechatInquiry.js [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+			emailNotifications: 'all',
+			answered: 'false',
+			// assignment: 'view'
+		};
+		RocketChat.models.Subscriptions.insert(subscriptionData);
+
+		// add user to room 
+		RocketChat.models.Rooms.addUsernameById(inquiry.rid, agent.username);
+
+		// TODO: mark inquiry as taken
```
Comment (rodrigok) [inline packages/rocketchat-livechat/server/models/LivechatInquiry.js]: Codacy Issue found: [Trailing spaces not allowed.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2726404280/issues/source?bid=3199634&fileBranchId=3348698#l45)

File: packages/rocketchat-livechat/server/models/LivechatInquiry.js [hunk 10]
```diff
@@ -0,0 +1,73 @@
+class LivechatInquiry extends RocketChat.models._Base {
+	constructor() {
+		super();
+		this._initModel('livechat_inquiry');
+
+		this.tryEnsureIndex({ 'rid': 1 }); // room id corresponding to this inquiry 
+		this.tryEnsureIndex({ 'name': 1 }); // name of the inquiry (client name for now)
+		this.tryEnsureIndex({ 'message': 1 }); // message sent by the client
```
Comment (rodrigok) [inline packages/rocketchat-livechat/server/models/LivechatInquiry.js]: Codacy Issue found: [Trailing spaces not allowed.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2726404280/issues/source?bid=3199634&fileBranchId=3348698#l8)

File: packages/rocketchat-livechat/client/views/sideNav/livechat.js [hunk 11]
```diff
@@ -36,6 +36,29 @@ Template.livechat.helpers({
...
+		var inqs = LivechatInquiry.find({
+			agents: user._id,
+			status: 'open'
+		}, {
+			sort: {
+				'ts' : 1
+			}
+		});
+
+		// for notification sound
```
Comment (rodrigok) [inline packages/rocketchat-livechat/client/views/sideNav/livechat.js]: Codacy Issue found: [Trailing spaces not allowed.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2726498730/issues/source?bid=3199634&fileBranchId=3348698#l52)

File: packages/rocketchat-livechat/client/views/sideNav/livechat.js [hunk 12]
```diff
@@ -36,6 +36,29 @@ Template.livechat.helpers({
 			}
 		});
 	},
+	inquiries() {
+		var user = Meteor.user();
+
+		// get all inquiries of the department
```
Comment (rodrigok) [inline packages/rocketchat-livechat/client/views/sideNav/livechat.js]: Codacy Issue found: [Trailing spaces not allowed.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2726498730/issues/source?bid=3199634&fileBranchId=3348698#l42)

File: packages/rocketchat-livechat/client/views/sideNav/livechat.js [hunk 13]
```diff
@@ -36,6 +36,29 @@ Template.livechat.helpers({
...
+
+		// for notification sound 
+		inqs.forEach(function (inq) {
+			KonchatNotification.newRoom(inq.rid);
+		});
+
+		return inqs;
+	},
+	guestPool() {
+		return RocketChat.settings.get('Livechat_Routing_Method') == 'Guest_Pool';
```
Comment (rodrigok) [inline packages/rocketchat-livechat/client/views/sideNav/livechat.js]: Codacy Issue found: [Expected '===' and instead saw '=='.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2726498730/issues/source?bid=3199634&fileBranchId=3348698#l60)

File: packages/rocketchat-livechat/client/views/sideNav/livechat.js [hunk 14]
```diff
@@ -36,6 +36,29 @@ Template.livechat.helpers({
...
+			agents: user._id,
+			status: 'open'
+		}, {
+			sort: {
+				'ts' : 1
+			}
+		});
+
+		// for notification sound 
+		inqs.forEach(function (inq) {
```
Comment (rodrigok) [inline packages/rocketchat-livechat/client/views/sideNav/livechat.js]: Codacy Issue found: [Unexpected space before function parentheses.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2726498730/issues/source?bid=3199634&fileBranchId=3348698#l53)

File: packages/rocketchat-livechat/client/views/sideNav/livechat.js [hunk 15]
```diff
@@ -62,3 +85,7 @@ Template.livechat.events({
 		});
 	}
 });
+
+Template.livechat.onCreated(function() {
+ 	this.subscribe('livechat:inquiry');
```
Comment (rodrigok) [inline packages/rocketchat-livechat/client/views/sideNav/livechat.js]: Codacy Issue found: [Mixed spaces and tabs.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2726498730/issues/source?bid=3199634&fileBranchId=3348698#l90)

File: packages/rocketchat-livechat/client/views/sideNav/livechat.js (hunk 13 above)
Comment (rodrigok) [inline packages/rocketchat-livechat/client/views/sideNav/livechat.js]: Codacy Issue found: [Expected '===' and instead saw '=='.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2726498730/issues/source?bid=3199634&fileBranchId=3348698#l60)

File: packages/rocketchat-ui-sidenav/side-nav/livechatInquiryItem.js [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+        var inquiry = this;
+
+        swal({
+            title: 'Do you want to take this client?',
+            text: 'message: ' + inquiry.message,
+            showCancelButton: true,
+            confirmButtonColor: '#3085d6',
+            cancelButtonColor: '#d33',
+            confirmButtonText: 'Take it!'
+        }, function (isConfirm) {
```
Comment (rodrigok) [inline packages/rocketchat-ui-sidenav/side-nav/livechatInquiryItem.js]: Codacy Issue found: [Unexpected space before function parentheses.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2726498881/issues/source?bid=3199634&fileBranchId=3348698#l49)

File: packages/rocketchat-livechat/server/models/LivechatInquiry.js [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+			emailNotifications: 'all',
+			answered: 'false',
+		};
+		RocketChat.models.Subscriptions.insert(subscriptionData);
+
+		// add user to room
+		RocketChat.models.Rooms.addUsernameById(inquiry.rid, agent.username);
+
+		// mark inquiry as taken
+		this.update({},{
```
Comment (rodrigok) [inline packages/rocketchat-livechat/server/models/LivechatInquiry.js]: Codacy Issue found: [A space is required after ','.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2726498771/issues/source?bid=3199634&fileBranchId=3348698#l45)

File: packages/rocketchat-livechat/config.js [hunk 18]
```diff
@@ -129,4 +129,15 @@ Meteor.startup(function() {
 		public: true,
 		i18nLabel: 'Apiai_Language'
 	});
+
+	RocketChat.settings.add('Livechat_Routing_Method', 'Least_Amount', {
```
Comment (rodrigok) [inline packages/rocketchat-livechat/config.js]: Codacy Issue found: [Trailing spaces not allowed.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2726498732/issues/source?bid=3199634&fileBranchId=3348698#l133)

File: packages/rocketchat-livechat/config.js [hunk 19]
```diff
@@ -129,4 +129,15 @@ Meteor.startup(function() {
 		public: true,
 		i18nLabel: 'Apiai_Language'
 	});
+
+	RocketChat.settings.add('Livechat_Routing_Method', 'Least_Amount', { 
+		type: 'select',
```
Comment (rodrigok) [inline packages/rocketchat-livechat/config.js]: Codacy Issue found: [Trailing spaces not allowed.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2726498732/issues/source?bid=3199634&fileBranchId=3348698#l134)

File: packages/rocketchat-livechat/server/lib/QueueMethods.js [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+			ts: new Date(),
+			v: {
+				_id: guest._id,
+				token: message.token,
+			},
+			open: true,
+		}, roomInfo);
+		RocketChat.models.LivechatInquiry.insert(inquiry);
+		RocketChat.models.Rooms.insert(room);
+
```
Comment (rodrigok) [inline packages/rocketchat-livechat/server/lib/QueueMethods.js]: Codacy Issue found: [Trailing spaces not allowed.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2726558665/issues/source?bid=3199634&fileBranchId=3348698#l111)

File: packages/rocketchat-livechat/server/lib/QueueMethods.js [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+		return room;
+	},
+	/* Guest Pool Queuing Method:
+	 *
+	 * An incomming livechat is created as an Inquiry 
+	 * which is picked up from an agent. 
+	 * An Inquiry is visible to all agents (TODO: in the correct department)
+     * 
+	 * A room is still created with the initial message, but it is occupied by
```
Comment (rodrigok) [inline packages/rocketchat-livechat/server/lib/QueueMethods.js]: Codacy Issue found: [Trailing spaces not allowed.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2726558665/issues/source?bid=3199634&fileBranchId=3348698#l63)

File: packages/rocketchat-livechat/server/lib/QueueMethods.js [hunk 22]
```diff
@@ -0,0 +1,114 @@
...
+	/* Guest Pool Queuing Method:
+	 *
+	 * An incomming livechat is created as an Inquiry 
+	 * which is picked up from an agent. 
+	 * An Inquiry is visible to all agents (TODO: in the correct department)
+     * 
+	 * A room is still created with the initial message, but it is occupied by 
+	 * only the client until paired with an agent 
+	 */
+	'Guest_Pool' : function (guest, message, roomInfo) {
```
Comment (rodrigok) [inline packages/rocketchat-livechat/server/lib/QueueMethods.js]: Codacy Issue found: [Unexpected space before function parentheses.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2726558665/issues/source?bid=3199634&fileBranchId=3348698#l66)

File: packages/rocketchat-livechat/server/lib/QueueMethods.js [hunk 23]
```diff
@@ -0,0 +1,114 @@
...
+		RocketChat.models.Subscriptions.insert(subscriptionData);
+
+		return room;
+	},
+	/* Guest Pool Queuing Method:
+	 *
+	 * An incomming livechat is created as an Inquiry 
+	 * which is picked up from an agent. 
+	 * An Inquiry is visible to all agents (TODO: in the correct department)
+     *
```
Comment (rodrigok) [inline packages/rocketchat-livechat/server/lib/QueueMethods.js]: Codacy Issue found: [Trailing spaces not allowed.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2726558665/issues/source?bid=3199634&fileBranchId=3348698#l62)

File: packages/rocketchat-livechat/server/lib/QueueMethods.js [hunk 24]
```diff
@@ -0,0 +1,114 @@
+RocketChat.QueueMethods = {
+	/* Least Amount Queuing method:
+	 *
+	 * default method where the agent with the least number
+	 * of open chats is paired with the incoming livechat
```
Comment (rodrigok) [inline packages/rocketchat-livechat/server/lib/QueueMethods.js]: Codacy Issue found: [Trailing spaces not allowed.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2726558665/issues/source?bid=3199634&fileBranchId=3348698#l5)

File: packages/rocketchat-livechat/server/lib/QueueMethods.js [hunk 25]
```diff
@@ -0,0 +1,114 @@
...
+		var inquiry = {
+			rid: message.rid,
+			message: message.msg,
+			name: guest.name || guest.username,
+			ts: new Date(),
+			code: roomCode,
+			department: guest.department,
+			agents: agentIds,
+			status: 'open'
+		}
```
Comment (rodrigok) [inline packages/rocketchat-livechat/server/lib/QueueMethods.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2726558665/issues/source?bid=3199634&fileBranchId=3348698#l93)

File: packages/rocketchat-livechat/server/lib/QueueMethods.js [hunk 26]
```diff
@@ -0,0 +1,114 @@
...
+		return room;
+	},
+	/* Guest Pool Queuing Method:
+	 *
+	 * An incomming livechat is created as an Inquiry 
+	 * which is picked up from an agent. 
+	 * An Inquiry is visible to all agents (TODO: in the correct department)
+     * 
+	 * A room is still created with the initial message, but it is occupied by 
+	 * only the client until paired with an agent
```
Comment (rodrigok) [inline packages/rocketchat-livechat/server/lib/QueueMethods.js]: Codacy Issue found: [Trailing spaces not allowed.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2726558665/issues/source?bid=3199634&fileBranchId=3348698#l64)

File: packages/rocketchat-livechat/server/lib/QueueMethods.js [hunk 27]
```diff
@@ -0,0 +1,114 @@
+RocketChat.QueueMethods = {
+	/* Least Amount Queuing method:
+	 *
+	 * default method where the agent with the least number
+	 * of open chats is paired with the incoming livechat 
+	 */
+	'Least_Amount' : function (guest, message, roomInfo) {
```
Comment (rodrigok) [inline packages/rocketchat-livechat/server/lib/QueueMethods.js]: Codacy Issue found: [Unexpected space before function parentheses.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2726558665/issues/source?bid=3199634&fileBranchId=3348698#l7)

File: packages/rocketchat-livechat/server/lib/QueueMethods.js [hunk 28]
```diff
@@ -0,0 +1,114 @@
...
+
+		RocketChat.models.Rooms.insert(room);
+		RocketChat.models.Subscriptions.insert(subscriptionData);
+
+		return room;
+	},
+	/* Guest Pool Queuing Method:
+	 *
+	 * An incomming livechat is created as an Inquiry 
+	 * which is picked up from an agent.
```
Comment (rodrigok) [inline packages/rocketchat-livechat/server/lib/QueueMethods.js]: Codacy Issue found: [Trailing spaces not allowed.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2726558665/issues/source?bid=3199634&fileBranchId=3348698#l60)

File: packages/rocketchat-livechat/server/lib/QueueMethods.js [hunk 29]
```diff
@@ -0,0 +1,114 @@
...
+		};
+
+		RocketChat.models.Rooms.insert(room);
+		RocketChat.models.Subscriptions.insert(subscriptionData);
+
+		return room;
+	},
+	/* Guest Pool Queuing Method:
+	 *
+	 * An incomming livechat is created as an Inquiry
```
Comment (rodrigok) [inline packages/rocketchat-livechat/server/lib/QueueMethods.js]: Codacy Issue found: [Trailing spaces not allowed.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2726558665/issues/source?bid=3199634&fileBranchId=3348698#l59)

File: packages/rocketchat-livechat/server/lib/QueueMethods.js [hunk 30]
```diff
@@ -0,0 +1,114 @@
...
+				token: message.token,
+			},
+			open: true,
+		}, roomInfo);
+		RocketChat.models.LivechatInquiry.insert(inquiry);
+		RocketChat.models.Rooms.insert(room);
+
+		return room;
+	}
+}
```
Comment (rodrigok) [inline packages/rocketchat-livechat/server/lib/QueueMethods.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2726640507/issues/source?bid=3199634&fileBranchId=3348698#l114)

File: packages/rocketchat-livechat/server/models/LivechatDepartmentAgents.js [hunk 31]
```diff
@@ -70,6 +70,34 @@ class LivechatDepartmentAgents extends RocketChat.models._Base {
...
+		var onlineUsernames = _.pluck(onlineUsers.fetch(), 'username');
+
+		var query = {
+			departmentId: departmentId,
+			username: {
+				$in: onlineUsernames
+			}
+		};
+
+		var agents = this.find(query);
```
Comment (rodrigok) [inline packages/rocketchat-livechat/server/models/LivechatDepartmentAgents.js]: Codacy Issue found: ["agents" is already defined](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2726640534/issues/source?bid=3199634&fileBranchId=3348698#l92)

File: packages/rocketchat-livechat/server/models/LivechatInquiry.js [hunk 32]
```diff
@@ -0,0 +1,44 @@
...
+		this.tryEnsureIndex({ 'agents': 1}); // Id's of the agents who can see the inquiry (handle departments)
+		this.tryEnsureIndex({ 'status': 1}); // 'open', 'taken'
+	}
+
+	/*
+	 * mark the inquiry as taken
+	 */
+	takeInquiry(inquiryId) {
+		this.update({
+			"_id": inquiryId
```
Comment (rodrigok) [inline packages/rocketchat-livechat/server/models/LivechatInquiry.js]: Codacy Issue found: [Strings must use singlequote.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2727001283/issues/source?bid=3199634&fileBranchId=3348698#l20)

Comment (rodrigok) [inline packages/rocketchat-livechat/server/models/LivechatInquiry.js]: Codacy Issue found: [Mixed double and single quotes.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2727001283/issues/source?bid=3199634&fileBranchId=3348698#l20)

**Comment by sampaiodiego** [comment]
wow. thank you @AlecTroemel for this amazing contribution.
I have tested it and have some considerations:

- by calling [Meteor.user() here](https://github.com/RocketChat/Rocket.Chat/pull/3507/files#diff-3859d46a0039cf6441f4e7ce3a961437R40) you're making the `inquiries` helper reactive to any user data (like status for example). if you want just the `_id` you can use `Meteor.userId()`
- multiple agents can take the same inquiry at the moment, what I think isn't desirable =)
- there is a [typo here](https://github.com/RocketChat/Rocket.Chat/pull/3507/files#diff-93086412abd5631ab21899e7ff30cbe8R7)
- there are some [missing translations](https://github.com/RocketChat/Rocket.Chat/pull/3507/files#diff-cbd4f486f0887f469ed8998ea18cad11R43)

I'm totally available to fix those issues.. I can give it some attention after today's release ok?

**Comment by AlecTroemel** [comment]
I can look at those issues right now if you haven't gotten to them already. For multiple agents being able to take an inquiry, is that possible in the UI (not that case on my local machine) or just no check in the takeInquiry method (I definitely don't check for that oops :0 )?

**Comment by sampaiodiego** [comment]
ok.. please do =) I have done nothing..

in the UI.. if an agent clicks on the incoming livechat entry but not accepts the inquiry immediatly, in this meantime another agent can quickly respond to it.. then if the first agent accepts it, both agents will be on the room =)

**Comment by AlecTroemel** [comment]
Oh gosh ya (the good old critical section problem).. hmm I could set an inquiry to a new status "grabbed" (or whatever) right when its clicked, then switch it back to "open" or to "taken" depending on the response in there response. That way you only see inquiries that you can safely take. That would however require new server methods.. and inquiries would possibly pop in and out (which would cause new sound pings) 

Could also do a check when an agent takes an inquiry to see if someone beat them to it, then give them another alert telling them the it was already taken. 

Also, for the typo you mentioned, were you referring to line 7 { publish: 'livechat:agents' } that should be { publish: 'livechat:inquiry' }?

**Comment by AlecTroemel** [comment]
Leaning towards the second option for the multi agent problem, (definitely the simplest)

File: packages/rocketchat-ui-sidenav/side-nav/livechatInquiryItem.js [hunk 33]
```diff
@@ -0,0 +1,65 @@
...
+            showCancelButton: true,
+            confirmButtonColor: '#3085d6',
+            cancelButtonColor: '#d33',
+            confirmButtonText: 'Take it!'
+        }, function(isConfirm) {
+            if (isConfirm) {
+                // make server method call
+                Meteor.call('livechat:takeInquiry', inquiry, Meteor.user(), function(error, result) {
+                    if (!error) {
+                        swal("success!")
```
Comment (rodrigok) [inline packages/rocketchat-ui-sidenav/side-nav/livechatInquiryItem.js]: Codacy Issue found: [Mixed double and single quotes.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2756771202/issues/source?bid=3199634&fileBranchId=3348698#l54)

File: packages/rocketchat-ui-sidenav/side-nav/livechatInquiryItem.js [hunk 34]
```diff
@@ -0,0 +1,65 @@
...
+Template.livechatInquiryItem.events({
+    'click .open-room': function(e) {
+        e.stopPropagation();
+        e.preventDefault();
+
+        var inquiry = this;
+
+        swal({
+            title: t('Livechat_Take_Confirm'),
+            text: t('Message') + ": " + inquiry.message,
```
Comment (rodrigok) [inline packages/rocketchat-ui-sidenav/side-nav/livechatInquiryItem.js]: Codacy Issue found: [Mixed double and single quotes.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2756771202/issues/source?bid=3199634&fileBranchId=3348698#l44)

File: packages/rocketchat-ui-sidenav/side-nav/livechatInquiryItem.js [hunk 35]
```diff
@@ -0,0 +1,65 @@
...
+                    }
+                });
+            }
+        });
+    },
+});
+
+Template.livechatInquiryItem.onDestroyed(function() {
+    swal.close();
+})
```
Comment (rodrigok) [inline packages/rocketchat-ui-sidenav/side-nav/livechatInquiryItem.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2756771202/issues/source?bid=3199634&fileBranchId=3348698#l65)

File: packages/rocketchat-ui-sidenav/side-nav/livechatInquiryItem.js (hunk 33 above)
Comment (rodrigok) [inline packages/rocketchat-ui-sidenav/side-nav/livechatInquiryItem.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2756771202/issues/source?bid=3199634&fileBranchId=3348698#l54)

Comment (rodrigok) [inline packages/rocketchat-ui-sidenav/side-nav/livechatInquiryItem.js]: Codacy Issue found: [Strings must use singlequote.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2756771202/issues/source?bid=3199634&fileBranchId=3348698#l54)

File: packages/rocketchat-ui-sidenav/side-nav/livechatInquiryItem.js (hunk 34 above)
Comment (rodrigok) [inline packages/rocketchat-ui-sidenav/side-nav/livechatInquiryItem.js]: Codacy Issue found: [Strings must use singlequote.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2756771202/issues/source?bid=3199634&fileBranchId=3348698#l44)

File: packages/rocketchat-livechat/server/methods/takeInquiry.js [hunk 36]
```diff
@@ -0,0 +1,43 @@
+Meteor.methods({
+	'livechat:takeInquiry'(inquiry, agent) {
+
+		console.log("this");
```
Comment (rodrigok) [inline packages/rocketchat-livechat/server/methods/takeInquiry.js]: Codacy Issue found: [Strings must use singlequote.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2756771086/issues/source?bid=3199634&fileBranchId=3348698#l4)

Comment (rodrigok) [inline packages/rocketchat-livechat/server/methods/takeInquiry.js]: Codacy Issue found: [Mixed double and single quotes.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2756771086/issues/source?bid=3199634&fileBranchId=3348698#l4)

File: packages/rocketchat-ui-sidenav/side-nav/livechatInquiryItem.js [hunk 37]
```diff
@@ -0,0 +1,64 @@
...
+Template.livechatInquiryItem.events({
+    'click .open-room': function(e) {
+        e.stopPropagation();
+        e.preventDefault();
+
+        var inquiry = this;
+
+        swal({
+            title: t('Livechat_Take_Confirm'),
+            text: t('Message') + ": " + inquiry.message,
```
Comment (rodrigok) [inline packages/rocketchat-ui-sidenav/side-nav/livechatInquiryItem.js]: Codacy Issue found: [Strings must use singlequote.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2756841210/issues/source?bid=3199634&fileBranchId=3348698#l44)

Comment (rodrigok) [inline packages/rocketchat-ui-sidenav/side-nav/livechatInquiryItem.js]: Codacy Issue found: [Mixed double and single quotes.](https://www.codacy.com/app/RocketChat/Rocket-Chat/file/2756841210/issues/source?bid=3199634&fileBranchId=3348698#l44)

**Comment by dhpowrhost** [comment]
How can i implement these changes?

**Comment by AlecTroemel** [comment]
@dhpowrhost if they merge this (given I didn't do anything wrong/bad!) you'll be able to pull from this repo. I also have a fork of rocket.chat with these changes implemented here https://github.com/AlecTroemel/Rocket.Chat on the custom branch.

**Comment by AlecTroemel** [comment]
scratch that the changes are the the **Develop** branch

**Comment by AlecTroemel** [comment]
changes made based on the comments above, also added the ability to return room as an inquiry (when the inquiry queueing method is chosen)
screen shot 2016-06-22 at 10 17 30 am
screen shot 2016-06-22 at 10 17 59 am

**Comment by sampaiodiego** [comment]
I just merged this at ea7483bcac3d56f07b6f4ce16a1942c39b9f942b

thanks again @AlecTroemel

**Pull request closed without merging**